File: app/src/main/java/com/github/overpass/gather/model/repo/login/AuthRepo.java
package com.github.overpass.gather.model.repo.login;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.github.overpass.gather.screen.auth.login.SignInStatus;
import com.google.firebase.auth.FirebaseAuth;

public class AuthRepo {

    private final FirebaseAuth auth;

    public AuthRepo(FirebaseAuth auth) {
        this.auth = auth;
    }

    public LiveData<SignInStatus> signIn(String email, String password) {
        MutableLiveData<SignInStatus> signInStatus = new MutableLiveData<>();
        signInStatus.setValue(new SignInStatus.Progress());
        auth.signInWithEmailAndPassword(email, password)
                .addOnSuccessListener(authResult -> {
                    signInStatus.setValue(new SignInStatus.Success());
                })
                .addOnFailureListener(e -> {
                    signInStatus.setValue(new SignInStatus.Error(e));
                });
        return signInStatus;
    }

    public void signOut(Runnable onSuccess) {
        auth.signOut();
        onSuccess.run();
    }
}

File: app/src/main/java/com/github/overpass/gather/model/repo/login/AuthRepo.kt
package com.github.overpass.gather.model.repo.login

import com.github.overpass.gather.model.commons.mapToSuccess
import com.github.overpass.gather.model.data.entity.signin.SignInStatus
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.FirebaseAuth
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AuthRepo @Inject constructor(private val auth: FirebaseAuth) {

    fun signIn(email: String, password: String): Task<SignInStatus> {
        return auth.signInWithEmailAndPassword(email, password)
                .mapToSuccess(
                        successMapper = { SignInStatus.Success },
                        failureMapper = { SignInStatus.Error(it) }
                )
    }

    fun signOut() {
        auth.signOut()
    }
}
